{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "sys.path.append(r'D:\\ChromatinTracingPipeline\\CommonTools')\n",
    "import MaxViewer as mv\n",
    "import IOTools as io\n",
    "import AlignmentTools as at\n",
    "\n",
    "DNA_data_folder = r'D:\\DNA_FISH\\E20200918_D0043\\DNA' #Folder with DNA dax files\n",
    "RNA_data_folder = r'D:\\DNA_FISH\\E20200918_D0043\\RNA' #Folder with RNA dax files\n",
    "Analysis_folder = r'D:\\DNA_FISH\\E20200918_D0043\\Analysis_CK' #Folder with analysis results of chromatin tracing\n",
    "os.makedirs(Analysis_folder+os.sep+\"RNA_DNA_Analysis\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob,os,pickle\n",
    "import numpy as np\n",
    "from collections import OrderedDict\n",
    "from collections import namedtuple\n",
    "\n",
    "#Get all folders in RNA data, excluding ones with an underscore (e.g. H0_postBleach)\n",
    "folders = [folder for folder in glob.glob(RNA_data_folder + os.sep + r'H*') if '_' not in os.path.basename(folder)]\n",
    "#Add the DAPI folder from the DNA data\n",
    "folders += glob.glob(DNA_data_folder + os.sep + r'H0*')\n",
    "#Get list of all FOVs, using 2nd folder in list. Assumes the same in all folders\n",
    "fovs = [os.path.basename(fl) for fl in glob.glob(folders[1]+'\\*.dax')]\n",
    "\n",
    "dic = pickle.load(open(Analysis_folder + os.sep + 'Selected_Spot.pkl','rb'))\n",
    "class_ids, coords, cell_ids = np.array(dic[u'class_ids']), np.array(dic[u'coords']), np.array(dic[u'cell_ids'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import namedtuple\n",
    "\n",
    "Vec3D = namedtuple('Vec3D', ['x', 'y', 'z'])\n",
    "DNA_Spot = namedtuple(\"DNA_Spot\", ['cell_id', 'x', 'y', 'z'])\n",
    "\n",
    "def norm(im,sz=40):\n",
    "    import cv2\n",
    "    im_ = im.astype(np.float32)\n",
    "    im_ = im_ / cv2.blur(im_,(sz,sz))\n",
    "    im_[im_>2] = 2\n",
    "    return im_\n",
    "\n",
    "def align(im1, im2):\n",
    "    \"\"\"Calculates the drift between two images\"\"\"\n",
    "    return at.fftalign_2d(norm(im1),norm(im2),max_disp=500)\n",
    "\n",
    "def prepare_data(ifov):\n",
    "    #Get list of dax files for a specific FOV (ifov) across all folders\n",
    "    files_ = [fld+os.sep+fovs[ifov] for fld in folders]\n",
    "    #Make sure the list is sorted with RNA first, in order, then DNA \n",
    "    #(probably not necessary since folders is already in the correct order?)\n",
    "    isort = np.argsort([int(os.path.basename(os.path.dirname(fl))[1:].split('R')[0].split('B')[0]) +  (-10 if 'RNA' in fl else 0)\n",
    "                        for fl in files_])\n",
    "    files_ = np.array(files_)[isort]\n",
    "\n",
    "    #Load the images for this FOV\n",
    "    ims = OrderedDict() #As of python 3.7, regular dicts have this functionality, but this is used for compatibility\n",
    "    for fl in files_:\n",
    "        if 'H0' in fl:\n",
    "            ncols = 5\n",
    "            cols = [(4, '_dapi')]\n",
    "        if 'H1' in fl:\n",
    "            ncols = 4\n",
    "            cols = [(0, '_mCherry'), (1, '_GFP'), (2, '_beads')]\n",
    "        if 'H2' in fl:\n",
    "            ncols = 4\n",
    "            cols = [(1, '_sox2'), (2, '_beads')]\n",
    "        for i, tag in cols:\n",
    "            name = os.path.basename(os.path.dirname(fl)) + tag\n",
    "            ims[name] = io.DaxReader(fl).loadAll()[i::ncols]\n",
    "\n",
    "    #Calculate the drift\n",
    "    imed = int(len(ims['H0_dapi']) / 2)\n",
    "    xtDNA, ytDNA = align(ims['H0_dapi'][imed], ims['H0B,B,B_dapi'][imed])\n",
    "    xtsox2, ytsox2 = align(ims['H1_beads'][imed], ims['H2_beads'][imed])\n",
    "    #print(f\"DAPI drift {ifov}: {xtDNA},{ytDNA}\")\n",
    "    \n",
    "    #Delete the bead images now that drift is calculated\n",
    "    ims.pop('H1_beads')\n",
    "    ims.pop('H2_beads')\n",
    "    \n",
    "    #Apply drift correction to the sox2 and DNA dapi images\n",
    "    ims['H2_sox2'] = at.translate(ims['H2_sox2'], [0,-xtsox2,-ytsox2])\n",
    "    ims['H0B,B,B_dapi'] = at.translate(ims['H0B,B,B_dapi'], [0,-xtDNA,-ytDNA])\n",
    "    \n",
    "    #Load the median images for 750 and cy5 colors and apply drift correction\n",
    "    #Note: These images are only used in visualization. May improve speed to skip\n",
    "    mean_dax1 = Analysis_folder + os.sep + os.path.basename(files_[1]).replace('.dax','_mean_0.dax')\n",
    "    mean_dax2 = Analysis_folder + os.sep + os.path.basename(files_[1]).replace('.dax','_mean_1.dax')\n",
    "    ims['DNA1'] = at.translate(io.DaxReader(mean_dax1).loadAll(),[0,-xtDNA,-ytDNA])\n",
    "    ims['DNA2'] = at.translate(io.DaxReader(mean_dax2).loadAll(),[0,-xtDNA,-ytDNA])\n",
    "    \n",
    "    #Get the DNA spot data for this FOV\n",
    "    spots_ = coords[class_ids==ifov]\n",
    "    cell_ids_ = cell_ids[class_ids==ifov]\n",
    "    x,y,z = spots_[:,0]+xtDNA, spots_[:,1]+ytDNA, spots_[:,2]\n",
    "    dna_spots = [DNA_Spot(cid, x_, y_, z_) for cid, x_, y_, z_ in zip(cell_ids_, x, y, z)]\n",
    "    #print(ifov, len(cell_ids_), len(dna_spots), Counter(Counter(cell_ids_).values()))\n",
    "    \n",
    "    return ims, dna_spots, (xtDNA,ytDNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pylab as plt\n",
    "\n",
    "def get_spot(image):\n",
    "    data = {}\n",
    "    \n",
    "    #Get coordinates of brightest pixel relative to center\n",
    "    pos = np.unravel_index(np.argmax(image),image.shape)-np.array(image.shape)/2.\n",
    "    #then L2 normalize (i.e. get distance of brightest pixel from center in nm)\n",
    "    data['dist'] = np.linalg.norm(pos*[200,108,108])\n",
    "    \n",
    "    #Max used as signal, median as noise\n",
    "    data['h'], data['bk'] = np.max(image), np.median(image)\n",
    "    data['ratio'] = data['h'] / data['bk']\n",
    "    \n",
    "    #Burst size (intensity) measured as sum of pixels minus background*pixel volume\n",
    "    data['intensity'] = np.sum(image) - data['bk']*image.size\n",
    "    \n",
    "    data['keep'] = int((data['ratio'] > 3) and (data['dist'] < 1000))\n",
    "    \n",
    "    data['im'] = np.max(image,0) #Will it break something if h is np.max(image,0)?\n",
    "    return data\n",
    "\n",
    "def subset(image, c, m):\n",
    "    return image[max(0,c.z-m.z):c.z+m.z, max(0,c.x-m.x):c.x+m.x, max(0,c.y-m.y):c.y+m.y]\n",
    "\n",
    "def find_RNA_spots(ims, dna_spots):\n",
    "    infos = []#OrderedDict()\n",
    "    keeps = []#OrderedDict()\n",
    "    plot = False\n",
    "    for ispot, dna_spot in enumerate(dna_spots):\n",
    "        try:\n",
    "            margin = Vec3D(x=10, y=10, z=6)\n",
    "            center = Vec3D(x=int(dna_spot.x), y=int(dna_spot.y), z=int(dna_spot.z))\n",
    "\n",
    "            sox2 = get_spot(subset(ims['H2_sox2'], center, margin))\n",
    "            mCherry = get_spot(subset(ims['H1_mCherry'], center, margin))\n",
    "            GFP = get_spot(subset(ims['H1_GFP'], center, margin))\n",
    "\n",
    "            infos.append([dna_spot, sox2, mCherry, GFP])\n",
    "            keeps.append([sox2['keep'], mCherry['keep'], GFP['keep']])\n",
    "\n",
    "            if plot:           \n",
    "                imDNA1 = np.max(subset(ims['DNA1'], center, margin), 0)\n",
    "                imDNA2 = np.max(subset(ims['DNA2'], center, margin), 0)\n",
    "                sy,sx,sz = 30,30,6\n",
    "                xc,yc,zc = int(dna_spot.x), int(dna_spot.y), int(dna_spot.z)\n",
    "                imDAPI1 = np.max(ims['H0_dapi'][zc:zc+1,xc-sx:xc+sx,yc-sy:yc+sy],0)\n",
    "                imDAPI2 = np.max(ims['H0B,B,B_dapi'][zc:zc+1,xc-sx:xc+sx,yc-sy:yc+sy],0)\n",
    "\n",
    "                fig,axs = plt.subplots(1,7,figsize=(10,3))\n",
    "                axs[0].imshow(imDNA1,cmap='gray')\n",
    "                axs[0].set_title('750')\n",
    "                axs[1].imshow(imDNA2,cmap='gray')\n",
    "                axs[1].set_title('Cy5')\n",
    "                axs[2].imshow(sox2['im'],cmap='gray')\n",
    "                axs[2].set_title('Sox2')\n",
    "                axs[3].imshow(mCherry['im'],cmap='gray')\n",
    "                axs[3].set_title('mCherry')\n",
    "                axs[4].imshow(GFP['im'],cmap='gray')\n",
    "                axs[4].set_title('GFP')\n",
    "                axs[5].imshow(imDAPI1,cmap='gray')\n",
    "                axs[5].set_title('RDAPI')\n",
    "                axs[6].imshow(imDAPI2,cmap='gray')\n",
    "                axs[6].set_title('DDAPI')\n",
    "                plt.suptitle('Spot '+str(ispot)+' cell:'+str(dna_spot.cell_id)+\\\n",
    "                             '\\nSox:'+str(keep_sox)+' '+str([np.round(H_sox,2),int(dist_sox)])+\\\n",
    "                            '  mCherry:'+str(keep_mcherry)+' '+str([np.round(H_mCherry,2),int(dist_mCherry)])+\\\n",
    "                            '  GFP:'+str(keep_GFP)+' '+str([np.round(H_GFP,2),int(dist_GFP)]))\n",
    "                plt.show()\n",
    "        except:\n",
    "            #print(dna_spot)\n",
    "            pass\n",
    "        \n",
    "    return infos, keeps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm_notebook as tqdm\n",
    "infos = []\n",
    "keeps = []\n",
    "drifts = []\n",
    "for ifov in tqdm(range(len(fovs))):\n",
    "    ims, dna_spots, driftxy = prepare_data(ifov)\n",
    "    drifts.append(driftxy)\n",
    "    infos_, keeps_ = find_RNA_spots(ims, dna_spots)\n",
    "    infos.extend(infos_)\n",
    "    keeps.extend(keeps_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "print(np.mean(np.array(keeps), axis=0))\n",
    "print(Counter([tuple(x) for x in keeps]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sox2h = [sox2['ratio'] for cid, sox2, gfp, mcherry in infos if sox2['keep'] and mcherry['keep'] and not gfp['keep']]\n",
    "mcherry2h = [mcherry['ratio'] for cid, sox2, gfp, mcherry in infos if sox2['keep'] and mcherry['keep'] and not gfp['keep']]\n",
    "plt.figure(figsize=(6,6))\n",
    "plt.title(\"{} spots\".format(len(sox2h)))\n",
    "plt.plot(sox2h,mcherry2h,'o')\n",
    "plt.xlabel('Sox2 signal-to-noise')\n",
    "plt.ylabel('mCherry signal-to-noise');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sox2h = [sox2['ratio'] for cid, sox2, gfp, mcherry in infos if sox2['keep'] and gfp['keep'] and not mcherry['keep']]\n",
    "gfp2h = [gfp['ratio'] for cid, sox2, gfp, mcherry in infos if sox2['keep'] and gfp['keep'] and not mcherry['keep']]\n",
    "plt.figure(figsize=(6,6))\n",
    "plt.title(\"{} spots\".format(len(sox2h)))\n",
    "plt.plot(sox2h,gfp2h,'o')\n",
    "plt.xlabel('Sox2 signal-to-noise')\n",
    "plt.ylabel('GFP signal-to-noise');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load single cell chromatin tracing data\n",
    "trace_data = {}\n",
    "with open(Analysis_folder + os.sep + \"Post_analysis\" + os.sep + \"single_cell_data.csv\") as f:\n",
    "    headers = f.readline().strip().split(',')\n",
    "    for line in f:\n",
    "        cols = line.strip().split(',')\n",
    "        cell_id = int(cols[6])\n",
    "        trace_data[cell_id] = {tag:float(num) for tag, num in zip(headers, cols)}\n",
    "        \n",
    "print(list(trace_data.values())[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cell_ids = [x[0].cell_id for x in infos]\n",
    "#print(Counter(cell_ids))\n",
    "#len(cell_ids)\n",
    "#print(Counter(class_ids))\n",
    "#ifov = 10\n",
    "#spots_ = coords[class_ids==ifov]\n",
    "#cell_ids_ = cell_ids[class_ids==ifov]\n",
    "#x,y,z = spots_[:,0], spots_[:,1], spots_[:,2]\n",
    "#dna_spots = [DNA_Spot(cid, x_, y_, z_) for cid, x_, y_, z_ in zip(cell_ids_, x, y, z)]\n",
    "#print(cell_ids_)\n",
    "#print(len(x))\n",
    "#print(dna_spots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "alldata = defaultdict(list)\n",
    "for dna_spot, sox2, mcherry, gfp in infos:\n",
    "    if dna_spot.cell_id in trace_data:\n",
    "        alldata[dna_spot.cell_id].append([dna_spot, sox2, mcherry, gfp])\n",
    "#for cell in alldata:\n",
    " #   for x in alldata[cell]:\n",
    "  #      print(cell, x[0])\n",
    "print(len(alldata))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cast_nums = []\n",
    "noncast_nums = []\n",
    "linked_data = []\n",
    "cast_dists = []\n",
    "noncast_dists = []\n",
    "missing = 0\n",
    "ambiguous = 0\n",
    "fovs = defaultdict(lambda: defaultdict(list))\n",
    "for dna_data in trace_data.values():\n",
    "    fov = int(dna_data['cell_id'] // 1000)\n",
    "    cast_coords = np.array([dna_data['CAST_y'], dna_data['CAST_x'], dna_data['CAST_z']])\n",
    "    noncast_coords = np.array([dna_data['129_y'], dna_data['129_x'], dna_data['129_z']])\n",
    "    fovs[fov]['precast'].append(cast_coords)\n",
    "    fovs[fov]['prenoncast'].append(noncast_coords)\n",
    "    cast_coords = np.array([dna_data['CAST_y']+drifts[fov][0]*109, dna_data['CAST_x']+drifts[fov][1]*109, dna_data['CAST_z']])\n",
    "    noncast_coords = np.array([dna_data['129_y']+drifts[fov][0]*109, dna_data['129_x']+drifts[fov][1]*109, dna_data['129_z']]) \n",
    "    cast_rna = None\n",
    "    cast_dist = 10000000\n",
    "    noncast_rna = None\n",
    "    noncast_dist = 10000000\n",
    "    fovs[fov]['cast'].append(cast_coords)\n",
    "    fovs[fov]['noncast'].append(noncast_coords)\n",
    "    for rna_data in alldata[dna_data['cell_id']]:\n",
    "        #num += f\"|{rna_data[1]['keep']},{rna_data[2]['keep']},{rna_data[3]['keep']}|\"\n",
    "        coords = np.array([rna_data[0].x*109, rna_data[0].y*109, rna_data[0].z*200])\n",
    "        fovs[fov]['rna'].append(coords)\n",
    "        castdist = np.linalg.norm(coords - cast_coords)\n",
    "        noncastdist = np.linalg.norm(coords - noncast_coords)\n",
    "        if castdist < cast_dist:\n",
    "            cast_rna = rna_data\n",
    "            cast_dist = castdist\n",
    "        if noncastdist < noncast_dist:\n",
    "            noncast_rna = rna_data\n",
    "            noncast_dist = noncastdist\n",
    "        #print()\n",
    "    if cast_rna is None or noncast_rna is None or cast_rna == noncast_rna:\n",
    "        #print(dna_data['cell_id'], drifts[fov][0]*109, drifts[fov][1]*109)\n",
    "        #print(cast_coords)\n",
    "        #print(noncast_coords)\n",
    "        #print([(rd[0].x*109, rd[0].y*109, rd[0].z*200) for rd in alldata[dna_data['cell_id']]])\n",
    "        #print()\n",
    "        if len(alldata[dna_data['cell_id']]) < 2:\n",
    "            missing += 1\n",
    "        else:\n",
    "            ambiguous += 1\n",
    "        continue\n",
    "    #nums.append(num)\n",
    "    cast_dists.append(cast_dist)\n",
    "    noncast_dists.append(noncast_dist)\n",
    "    cast_nums.append(f\"{cast_rna[1]['keep']},{cast_rna[2]['keep']},{cast_rna[3]['keep']}\")\n",
    "    noncast_nums.append(f\"{noncast_rna[1]['keep']},{noncast_rna[2]['keep']},{noncast_rna[3]['keep']}\")\n",
    "    linked_data.append({'cast_rna': cast_rna, '129_rna': noncast_rna, 'dna': dna_data})\n",
    "    \n",
    "print(len(linked_data))\n",
    "print(len(trace_data))\n",
    "print(missing, ambiguous)\n",
    "print(Counter(cast_nums))\n",
    "print(Counter(noncast_nums))\n",
    "\n",
    "print(np.median(cast_dists+noncast_dists))\n",
    "#import seaborn as sns\n",
    "#sns.histplot(cast_dists+noncast_dists)\n",
    "                        \n",
    "#plt.figure(figsize=(10,10))\n",
    "#ifov=10\n",
    "#precastx = [p[0] for p in fovs[ifov]['precast']]\n",
    "#precasty = [p[1] for p in fovs[ifov]['precast']]\n",
    "#prenoncastx = [p[0] for p in fovs[ifov]['prenoncast']]\n",
    "#prenoncasty = [p[1] for p in fovs[ifov]['prenoncast']]\n",
    "#castx = [p[0] for p in fovs[ifov]['cast']]\n",
    "#casty = [p[1] for p in fovs[ifov]['cast']]\n",
    "#noncastx = [p[0] for p in fovs[ifov]['noncast']]\n",
    "#noncasty = [p[1] for p in fovs[ifov]['noncast']]\n",
    "#rnax = [p[0] for p in fovs[ifov]['rna']]\n",
    "#rnay = [p[1] for p in fovs[ifov]['rna']]\n",
    "#plt.scatter(precastx, precasty, color=\"#cc4444\")\n",
    "#plt.scatter(prenoncastx, prenoncasty, color=\"#44cc44\")\n",
    "#plt.scatter(castx, casty, color=\"#ff0000\")\n",
    "#plt.scatter(noncastx, noncasty, color=\"#00ff00\")\n",
    "#plt.scatter(rnax, rnay, color=\"#0000ff\", alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import upsetplot\n",
    "\n",
    "def convert_nums(nums):\n",
    "    memberships = defaultdict(int)\n",
    "    for num in nums:\n",
    "        sox2, mch, gfp = num.split(',')\n",
    "        labels = []\n",
    "        if sox2 == '1':\n",
    "            labels.append('Sox2')\n",
    "        if mch == '1':\n",
    "            labels.append(' mCherry')\n",
    "        if gfp == '1':\n",
    "            labels.append(' GFP')\n",
    "        memberships[tuple(labels)] += 1\n",
    "    return memberships\n",
    "\n",
    "castmem = convert_nums(cast_nums)\n",
    "noncastmem = convert_nums(noncast_nums)\n",
    "print(castmem.keys())\n",
    "keys = [(), ('Sox2',), ('Sox2', ' mCherry'), ('Sox2', ' GFP'), (' mCherry',), (' GFP',), (' mCherry', ' GFP'), ('Sox2', ' mCherry', ' GFP')]\n",
    "#keys = [('Sox2',), ('mCherry',), ('GFP',)]\n",
    "dfc = upsetplot.from_memberships(keys, data=[castmem[k] for k in keys])\n",
    "dfnc = upsetplot.from_memberships(keys, data=[noncastmem[k] for k in keys])\n",
    "upsetplot.plot(dfc, sort_by='cardinality', sort_categories_by=None, show_counts=True)\n",
    "plt.savefig(Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+\"cast_categories_plot_bogdan.png\",dpi=300)\n",
    "upsetplot.plot(dfnc, sort_by='cardinality', sort_categories_by=None, show_counts=True)\n",
    "plt.savefig(Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+\"129_categories_plot_bogdan.png\",dpi=300)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save results to csv\n",
    "with open(Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+\"binary_celldata.csv\", \"w\") as f:\n",
    "    for data in linked_data:\n",
    "        line = [int(data['dna']['cell_id'])]\n",
    "        line.extend([data['cast_rna'][i]['keep'] for i in [1,2,3]])\n",
    "        line.extend([data['129_rna'][i]['keep'] for i in [1,2,3]])\n",
    "        print(','.join([str(x) for x in line]), file=f)\n",
    "    \n",
    "newdata = {}\n",
    "with open(r\"C:\\Users\\ckern\\Downloads\\rna2alleles_d43.csv\") as f:\n",
    "    header = f.readline().strip()\n",
    "    for line in f:\n",
    "        cols = line.strip().split(',')\n",
    "        newdata[int(cols[0])] = line.strip()\n",
    "       \n",
    "for data in linked_data:\n",
    "    cell = int(data['dna']['cell_id'])\n",
    "    newdata[cell] += ','+','.join([str(data['cast_rna'][i]['keep']) for i in [1,2,3]])\n",
    "    newdata[cell] += ','+','.join([str(data['129_rna'][i]['keep']) for i in [1,2,3]])\n",
    "    \n",
    "header += \",CAST_Sox2,CAST_mCherry,CAST_GFP,129_Sox2,129_mCherry,129_GFP\"\n",
    "with open(r\"C:\\Users\\ckern\\Downloads\\rna2alleles_d43.csv\", \"w\") as f:\n",
    "    print(header,file=f)\n",
    "    for cell, line in newdata.items():\n",
    "        print(line, file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "pickle.dump(linked_data, open(os.path.join(Analysis_folder, \"RNA_DNA_Analysis\", \"pickled_data.pkl\"), 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "linked_data = pickle.load(open(os.path.join(Analysis_folder, \"RNA_DNA_Analysis\", \"pickled_data.pkl\"), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def histo_plot(datakey, xlabel):\n",
    "    bursting = []\n",
    "    silent = []\n",
    "    cast_burst = []\n",
    "    cast_silent = []\n",
    "    noncast_burst = []\n",
    "    noncast_silent = []\n",
    "    for data in linked_data:\n",
    "        if (data['cast_rna'][1]['keep'] or data['cast_rna'][3]['keep']) and not data['cast_rna'][2]['keep']:\n",
    "        #if data['cast_rna'][1]['keep'] and data['cast_rna'][3]['keep']:\n",
    "            bursting.append(data['dna']['CAST_'+datakey])\n",
    "            cast_burst.append(data['dna']['CAST_'+datakey])\n",
    "        elif not data['cast_rna'][1]['keep'] and not data['cast_rna'][2]['keep'] and not data['cast_rna'][3]['keep']:\n",
    "        #elif not data['cast_rna'][1]['keep'] and not data['cast_rna'][3]['keep']:\n",
    "            silent.append(data['dna']['CAST_'+datakey])\n",
    "            cast_silent.append(data['dna']['CAST_'+datakey])\n",
    "        if (data['129_rna'][1]['keep'] or data['129_rna'][2]['keep']) and not data['129_rna'][3]['keep']:\n",
    "        #if data['129_rna'][1]['keep'] and data['129_rna'][2]['keep']:\n",
    "            bursting.append(data['dna']['129_'+datakey])\n",
    "            noncast_burst.append(data['dna']['129_'+datakey])\n",
    "        elif not data['129_rna'][1]['keep'] and not data['129_rna'][2]['keep'] and not data['129_rna'][3]['keep']:\n",
    "        #elif not data['129_rna'][1]['keep'] and not data['129_rna'][2]['keep']:\n",
    "            silent.append(data['dna']['129_'+datakey])\n",
    "            noncast_silent.append(data['dna']['129_'+datakey])\n",
    "\n",
    "    bursting = [x for x in bursting if not math.isnan(x)]\n",
    "    silent = [x for x in silent if not math.isnan(x)]\n",
    "    cast_burst = [x for x in cast_burst if not math.isnan(x)]\n",
    "    cast_silent = [x for x in cast_silent if not math.isnan(x)]\n",
    "    noncast_burst = [x for x in noncast_burst if not math.isnan(x)]\n",
    "    noncast_silent = [x for x in noncast_silent if not math.isnan(x)]\n",
    "    import seaborn\n",
    "    from scipy.stats import chisquare, ranksums, mannwhitneyu\n",
    "    def make_plot(bst, slnt, fileadd=''):\n",
    "        bstchi = [len([b for b in bst if b <= 250]), len([b for b in bst if b > 250])]\n",
    "        sltchi = [len([b for b in slnt if b <= 250]), len([b for b in slnt if b > 250])]\n",
    "        plt.figure(figsize=(6,5))\n",
    "        #seaborn.distplot(bst, label=f\"Bursting (n={len(bst)};mean={sum(bst)/len(bst):.0f}nm)\", kde=False)\n",
    "        #seaborn.distplot(slnt, label=f\"Resting (n={len(slnt)};mean={sum(slnt)/len(slnt):.0f}nm)\", kde=False)\n",
    "        seaborn.distplot(bst, label=f\"CAST (n={len(bst)};mean={sum(bst)/len(bst):.0f}nm)\", kde=False)\n",
    "        seaborn.distplot(slnt, label=f\"129 (n={len(slnt)};mean={sum(slnt)/len(slnt):.0f}nm)\", kde=False)\n",
    "        plt.xlabel(xlabel)\n",
    "        plt.ylabel(\"Chromosomes\")\n",
    "        #title =f\"Bursting: {100*bstchi[0]/(bstchi[0]+bstchi[1]):.1f}% in contact ({bstchi[0]}/{bstchi[1]})\\n\"\n",
    "        #title +=f\"Resting: {100*sltchi[0]/(sltchi[0]+sltchi[1]):.1f}% in contact ({sltchi[0]}/{sltchi[1]})\\n\"\n",
    "        #title += f\"Chi-square p-value={chisquare(bstchi, sltchi)[1]:0.5f}\"\n",
    "        title = f\"Wilcoxon p-value={ranksums(bst, slnt)[1]:0.5f}\\n\"\n",
    "        title += f\"Mann-Whitney p-value={mannwhitneyu(bst, slnt)[1]:0.5f}\"\n",
    "        plt.title(title)\n",
    "        #plt.axvline(250, linestyle=\"dashed\", color='red',label='Contact threshold: 250nm')\n",
    "        plt.legend()\n",
    "        plt.savefig(Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+\"rna_histogram_\"+fileadd+datakey+\".png\",dpi=300)\n",
    "    #make_plot(bursting, silent)\n",
    "    #make_plot(cast_burst, cast_silent, 'CAST_')\n",
    "    #make_plot(noncast_burst, noncast_silent, '129_')\n",
    "    make_plot(cast_burst, noncast_burst, 'burst_')\n",
    "    make_plot(cast_silent, noncast_silent, 'silent_')\n",
    "\n",
    "#histo_plot('ep_dist', 'E-P distance')\n",
    "#histo_plot('ins', 'Insulation (10-25, 25-33)')#\n",
    "#histo_plot('rgs_10_25', 'Radius of gyration (10-25)')\n",
    "#histo_plot('rgs_10_33', 'Radius of gyration (10-33)')\n",
    "#histo_plot('rgs_1_42', 'Radius of gyration (1-42)')\n",
    "#histo_plot('rgs_10_39', 'Radius of gyration (10-39)')\n",
    "#histo_plot('rgs_25_33', 'Radius of gyration (25-33)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "from scipy.stats import chisquare\n",
    "cast_burst = []\n",
    "noncast_burst = []\n",
    "cast_silent = []\n",
    "noncast_silent = []\n",
    "for data in linked_data:\n",
    "    if (data['cast_rna'][1]['keep'] or data['cast_rna'][3]['keep']) and not data['cast_rna'][2]['keep']:\n",
    "        cast_burst.append(data['dna']['CAST_ep_dist'])\n",
    "    elif not data['cast_rna'][1]['keep'] and not data['cast_rna'][2]['keep'] and not data['cast_rna'][3]['keep']:\n",
    "        cast_silent.append(data['dna']['CAST_ep_dist'])\n",
    "    if (data['129_rna'][1]['keep'] or data['129_rna'][2]['keep']) and not data['129_rna'][3]['keep']:\n",
    "        noncast_burst.append(data['dna']['129_ep_dist'])\n",
    "    elif not data['129_rna'][1]['keep'] and not data['129_rna'][2]['keep'] and not data['129_rna'][3]['keep']:\n",
    "        noncast_silent.append(data['dna']['129_ep_dist'])\n",
    "cast_burst = [x for x in cast_burst if not math.isnan(x)]\n",
    "cast_silent = [x for x in cast_silent if not math.isnan(x)]\n",
    "noncast_burst = [x for x in noncast_burst if not math.isnan(x)]\n",
    "noncast_silent = [x for x in noncast_silent if not math.isnan(x)]\n",
    "bursting = list(sorted(cast_burst + noncast_burst))\n",
    "silent = list(sorted(cast_silent + noncast_silent))\n",
    "\n",
    "from scipy.stats import chi2_contingency\n",
    "def get_pvalue(bursting, silent, thresh):\n",
    "    bstchi = [len([b for b in bursting if b <= thresh]), len([b for b in bursting if b > thresh])]\n",
    "    sltchi = [len([b for b in silent if b <= thresh]), len([b for b in silent if b > thresh])]\n",
    "    if 0 in bstchi or 0 in sltchi:\n",
    "        return 1\n",
    "    return chi2_contingency([bstchi, sltchi])[1]\n",
    "    \n",
    "def cumulative_graph(bursting, silent, name1, name2):\n",
    "    pvals = [get_pvalue(bursting, silent, t) for t in range(0,1500,20)]\n",
    "    bstchi = [len([b for b in bursting if b <= 250]), len([b for b in bursting if b > 250])]\n",
    "    sltchi = [len([b for b in silent if b <= 250]), len([b for b in silent if b > 250])]\n",
    "    plt.figure(figsize=(6,5))\n",
    "    for i, pval in enumerate(pvals):\n",
    "        if pval <= 0.01:\n",
    "            plt.axvspan((20*i)-10, (20*i)+10, color='#bbbbbb')\n",
    "        elif pval <= 0.05:\n",
    "            plt.axvspan((20*i)-10, (20*i)+10, color='#dddddd')\n",
    "    plt.plot(bursting, [(x+1)/len(bursting) for x in range(len(bursting))], label=name1)\n",
    "    plt.plot(silent, [(x+1)/len(silent) for x in range(len(silent))], label=name2)\n",
    "    plt.xlabel(\"E-P distance (nm)\")\n",
    "    plt.ylabel(\"Fraction of chromosomes\")\n",
    "    title = f\"{name1}: {100*bstchi[0]/(bstchi[0]+bstchi[1]):.1f}% in contact ({bstchi[0]}/{bstchi[1]})\\n\"\n",
    "    title += f\"{name2}: {100*sltchi[0]/(sltchi[0]+sltchi[1]):.1f}% in contact ({sltchi[0]}/{sltchi[1]})\\n\"\n",
    "    title += f\"p-value={chisquare(bstchi, sltchi)[1]:0.5f}, \"\n",
    "    title += f\"{chi2_contingency([bstchi, sltchi])[1]:0.5f}\"\n",
    "    plt.title(title)\n",
    "    plt.axvline(250, linestyle=\"dashed\", color='red',label='Contact threshold: 250nm')\n",
    "    plt.legend()\n",
    "    #plt.savefig(Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+f\"cumulative_ep_dist_{name1}_{name2}_d43.png\",dpi=300)\n",
    "    #plt.savefig(Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+f\"cumulative_ep_dist_{name1}_{name2}_d43.pdf\",dpi=300)\n",
    "\n",
    "cumulative_graph(bursting, silent, \"Bursting\", \"Resting\")\n",
    "cumulative_graph(list(sorted(cast_burst+cast_silent)), list(sorted(noncast_burst+noncast_silent)), \"CAST\", \"129\")\n",
    "cumulative_graph(list(sorted(cast_burst)), list(sorted(cast_silent)), \"CAST bursting\", \"CAST resting\")\n",
    "cumulative_graph(list(sorted(noncast_burst)), list(sorted(noncast_silent)), \"129 bursting\", \"129 resting\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import seaborn as sns\n",
    "import math\n",
    "\n",
    "def bursting(rna_data, i, j):\n",
    "    return (rna_data[1]['keep'] or rna_data[i]['keep']) and not rna_data[j]['keep']\n",
    "\n",
    "def resting(rna_data):\n",
    "    return not rna_data[1]['keep'] and not rna_data[2]['keep'] and not rna_data[3]['keep']\n",
    "\n",
    "def ratio(threshold, nums1, nums2):\n",
    "    n1_ratio = len([n for n in nums1 if n <= threshold]) / len(nums1)\n",
    "    n2_ratio = len([n for n in nums2 if n <= threshold]) / len(nums2)\n",
    "    return n1_ratio / n2_ratio\n",
    "\n",
    "def threshold_plot(dists, group1, group2, name1, name2):\n",
    "    ratio_nums = [math.log2(ratio(thresh, group1, group2)) for thresh in dists]\n",
    "    plt.figure()\n",
    "    plt.xlabel(\"E-P contact threshold (nm)\")\n",
    "    plt.ylabel(f\"Ratio of contact frequency ({name1}/{name2})\")\n",
    "    sns.lineplot(dists, ratio_nums)\n",
    "    plt.ylim(-1,1)\n",
    "    #plt.yticks([0.5,.75,1.0,1.25,1.5])\n",
    "    plt.xticks([150,250,350,450,550,650,750])\n",
    "    plt.tight_layout()\n",
    "    filename = Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+\"thresholdplot_\"+name1+\"_\"+name2\n",
    "    #plt.savefig(filename+\".png\",dpi=300)\n",
    "    \n",
    "dists = list(range(150,750,25))\n",
    "bursting_cast = [data['dna']['CAST_ep_dist'] for data in linked_data if bursting(data['cast_rna'], 3, 2)]\n",
    "resting_cast = [data['dna']['CAST_ep_dist'] for data in linked_data if resting(data['cast_rna'])]\n",
    "bursting_129 = [data['dna']['129_ep_dist'] for data in linked_data if bursting(data['129_rna'], 2, 3)]\n",
    "resting_129 = [data['dna']['129_ep_dist'] for data in linked_data if resting(data['129_rna'])]\n",
    "bursting_all = bursting_cast + bursting_129\n",
    "resting_all = resting_cast + resting_129\n",
    "all_cast = bursting_cast + resting_cast\n",
    "all_129 = bursting_129 + resting_129\n",
    "\n",
    "threshold_plot(dists, all_cast, all_129, \"CAST\", \"129\")\n",
    "threshold_plot(dists, resting_all, bursting_all, \"Resting\", \"Bursting\")\n",
    "threshold_plot(dists, resting_cast, bursting_cast, \"CAST Resting\", \"CAST Bursting\")\n",
    "threshold_plot(dists, resting_129, bursting_129, \"129 Resting\", \"129 Bursting\")\n",
    "threshold_plot(dists, bursting_cast, bursting_129, \"CAST Bursting\", \"129 Bursting\")\n",
    "threshold_plot(dists, resting_cast, resting_129, \"CAST Resting\", \"129 Resting\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "linked_data = pickle.load(open('D:\\DNA_FISH\\E20200918_D0043\\Analysis_CK\\RNA_DNA_Analysis\\linked_data.pkl', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "from scipy.stats import sem,ks_2samp\n",
    "\n",
    "def bursting(rna_data, i, j):\n",
    "    return (rna_data[1]['keep'] or rna_data[i]['keep']) and not rna_data[j]['keep']\n",
    "\n",
    "def resting(rna_data):\n",
    "    return not rna_data[1]['keep'] and not rna_data[2]['keep'] and not rna_data[3]['keep']\n",
    "\n",
    "def ratio(threshold, nums1, nums2):\n",
    "    n1_ratio = len([n for n in nums1 if n <= threshold]) / len(nums1)\n",
    "    n2_ratio = len([n for n in nums2 if n <= threshold]) / len(nums2)\n",
    "    return n1_ratio / n2_ratio\n",
    "\n",
    "\n",
    "dists = list(range(150,750,25))\n",
    "bursting_cast = [data['dna']['CAST_ep_dist'] for data in linked_data if bursting(data['cast_rna'], 3, 2)]\n",
    "resting_cast = [data['dna']['CAST_ep_dist'] for data in linked_data if resting(data['cast_rna'])]\n",
    "bursting_129 = [data['dna']['129_ep_dist'] for data in linked_data if bursting(data['129_rna'], 2, 3)]\n",
    "resting_129 = [data['dna']['129_ep_dist'] for data in linked_data if resting(data['129_rna'])]\n",
    "bursting_all = bursting_cast + bursting_129\n",
    "\n",
    "def ratio(threshold, nums1, nums2):\n",
    "    l1 = len([x for x in nums1 if x <= threshold])\n",
    "    l2 = len([x for x in nums2 if x <= threshold])\n",
    "    ratio = l1 / (l1+l2)\n",
    "    err = 1.96 * math.sqrt((ratio*(1-ratio)) / (l1+l2))\n",
    "    return ratio, err\n",
    "\n",
    "\n",
    "xticks = range(150,751,25)\n",
    "n1 = [ratio(x, bursting_cast, resting_cast) for x in xticks]\n",
    "n2 = [ratio(x, bursting_129, resting_129) for x in xticks]\n",
    "nums1 = [n[0]*100 for n in n1]\n",
    "err1 = [n[1]*100 for n in n1]\n",
    "nums2 = [n[0]*100 for n in n2]\n",
    "err2 = [n[1]*100 for n in n2]\n",
    "print(ks_2samp(nums1, nums2))\n",
    "plt.figure()\n",
    "plt.xlabel(\"E-P contact threshold (nm)\")\n",
    "plt.ylabel(f\"Fraction of alleles bursting (%)\")\n",
    "#plt.errorbar(xticks, nums1, yerr=err1, label=\"CAST\")\n",
    "#plt.errorbar(xticks, nums2, yerr=err2, label=\"129\")\n",
    "plt.plot(xticks, nums1, label=\"CAST\")\n",
    "plt.plot(xticks, nums2, label=\"129\")\n",
    "plt.fill_between(xticks, [x-y for x,y in zip(nums1,err1)], [x+y for x,y in zip(nums1,err1)], alpha=0.5)\n",
    "plt.fill_between(xticks, [x-y for x,y in zip(nums2,err2)], [x+y for x,y in zip(nums2,err2)], alpha=0.5)\n",
    "\n",
    "#plt.fill_between(xticks, nums2-err2, nums2+err2, alpha=0.5)\n",
    "#plt.ylim(0.5,1.5)\n",
    "#plt.yticks([0.5,.75,1.0,1.25,1.5])\n",
    "plt.xticks([150,300,450,600,750])\n",
    "#plt.ylim(ymin=0.2, ymax=0.8)\n",
    "#plt.yticks([0.2,0.3,0.4,0.5,0.6,0.7,0.8])\n",
    "plt.ylim(ymin=20, ymax=80)\n",
    "plt.yticks([20,30,40,50,60,70,80])\n",
    "plt.legend()\n",
    "plt.tight_layout()\n",
    "import os\n",
    "Analysis_folder = r'D:\\DNA_FISH\\E20200918_D0043\\Analysis_CK'\n",
    "filename = Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+\"bursting_fraction_d43\"\n",
    "plt.savefig(filename+\".png\",dpi=300)\n",
    "plt.savefig(filename+\".pdf\",dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import seaborn\n",
    "from scipy import stats\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "def pearson(x, y):\n",
    "    return stats.pearsonr(x, y)\n",
    "\n",
    "def scatter_plot(xdatakey, xlabel, ydatakey, ylabel, cast=True, noncast=True):\n",
    "    x=[]\n",
    "    y=[]\n",
    "    for data in linked_data:\n",
    "        if cast:\n",
    "            x.append(data['cast_rna'][1][xdatakey])\n",
    "            y.append(data['dna']['CAST_'+ydatakey])\n",
    "        if noncast:\n",
    "            x.append(data['129_rna'][1][xdatakey])\n",
    "            y.append(data['dna']['129_'+ydatakey])\n",
    "    #seaborn.regplot(x,y)\n",
    "    seaborn.jointplot(x, y, kind=\"reg\", stat_func=pearson)       \n",
    "    plt.xlabel(xlabel)\n",
    "    plt.ylabel(ylabel)\n",
    "    plt.tight_layout()\n",
    "    filename = Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+\"scatterplot_\"+xdatakey+\"_v_\"+ydatakey\n",
    "    if cast and not noncast:\n",
    "        filename += '_CAST'\n",
    "    if noncast and not cast:\n",
    "        filename += '_129'\n",
    "    plt.savefig(filename+\".png\",dpi=300)\n",
    "    \n",
    "xaxes = [('h', 'Sox2 brightness'), ('ratio', 'Sox2 signal-to-noise'), ('intensity', 'Sox2 intensity')]\n",
    "yaxes = [('ep_dist', 'E-P distance'), ('ins', 'Insulation (10-25, 25-33)'), ('rgs_10_25', 'Radius of gyration (10-25)'),\n",
    "         ('rgs_10_33', 'Radius of gyration (10-33)'), ('rgs_1_42', 'Radius of gyration (1-42)'),\n",
    "         ('rgs_10_39', 'Radius of gyration (10-39)'), ('rgs_25_33', 'Radius of gyration (25-33)')]\n",
    "for xdatakey, xlabel in xaxes:\n",
    "    for ydatakey, ylabel in yaxes:\n",
    "        scatter_plot(xdatakey, xlabel, ydatakey, ylabel, cast=True, noncast=True)\n",
    "        scatter_plot(xdatakey, xlabel, ydatakey, ylabel, cast=True, noncast=False)\n",
    "        scatter_plot(xdatakey, xlabel, ydatakey, ylabel, cast=False, noncast=True)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the data for input to chromatin tracing pipeline\n",
    "pickle.dump(linked_data, open(Analysis_folder+os.sep+\"RNA_DNA_Analysis\"+os.sep+\"linked_data.pkl\", 'wb'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
